### DISCUSSION TITLE
[QUESTION] How to get actual frame data with Dropped frames

### DISCUSSION THREAD

**Discussion by shritib** [awslabs/amazon-kinesis-video-streams-producer-c#390] [Q&A] [answered]
We are using KVS Producer C SDK to ingest video from camera to KVS stream. We want to handle network failure scenario in a way that we can persist frames on file system which are dropped due to bad network and retry it later using Oflline KVS Client. For this, we are looking to override callbacks provided by SDK, especially [droppedFrameReportFn](https://github.com/awslabs/amazon-kinesis-video-streams-producer-c/blob/1d9f34d83f84406159e2045bdd7ffa5aa2669971/src/include/com/amazonaws/kinesis/video/cproducer/Include.h#L208). This function returns only timecode of dropped [frame](https://github.com/awslabs/amazon-kinesis-video-streams-producer-c/blob/1d9f34d83f84406159e2045bdd7ffa5aa2669971/src/source/StreamCallbacksProvider.c#L38). 

One way we can achieve is keep frames in memory and when dropped frame callback is received from KVS, we take that frame data and persist on file system on camera. Since, we have limited RAM on camera and we know that KVS already persists frames in ContentView for buffer duration. And generates dropped frame callback only when buffer duration is passed and frame is not being sent to backend. 

So, Is it some way possible to get frame data along with timecode in dropped frame callback from KVS ?

**Comment by hassanctech** [comment, marked as answer]
Interesting.  At a high level I see a couple of different options here.  We could make a copy of the buffer here and pass it to you, because the callback cannot give you a pointer to what is in the content view its stored in the same content store that all streams store video into and access to it is guarded by a mutex so that's not possible.  We offer memory modes where all allocations from the SDK occur from a single allocation performed at the start, so we don't malloc all over the place, so there is no way you can get access to that so a memcpy is unavoidable. But here you are increasing the RAM now while we hand it off to your application which would be responsible for freeing it.  An alternative is we could accept a path where we write it out give a output handle (could be to file for example) when it fails.  I do not think the dropped frame report fn callback is the appropriate place for this as the name suggests it's just report/notification and the reason why it doesn't give you the frame data is because it would need to allocate the memory to do so.

Today there is no way to achieve this.  We have do have env variable that you can export `KVS_DEBUG_DUMP_DATA_FILE_DIR` if it's set it will write to that file whatever was sent to KVS backend, so this will not give you want to want since you're looking for the frames that didn't get sent.  But thought I'd mention it in case it's useful.
